### ISSUE TITLE
MPD/ffmpeg noise issue 

### ISSUE THREAD

**Issue by janui** [MusicPlayerDaemon/MPD#380]
The missing part of my setup is the integration of digital radio from my SAT-TV receiver. My first try was to use DLNA to access MPEG transport streams using the webradio feature of RuneAudio. This was successful until the upgrade of MPD to the actual release (0.20.21) because ffmpeg has been upgraded to 4.0.2 also. And ffmpeg 4.0.2 changed the sample format detected in the mp2-stream as fltp instead of s16p as it did earlier. Consquence: The ffmpeg plugin invoked by mpd performs a (meaningless) conversion from 48000:f:2 to 48000:32:2 producing noise.

Will also post on the ffmpeg forum

**Comment by MaxKellermann** [comment]
https://www.musicpd.org/help/
How to reproduce?
What kind of noise is this? Can you capture it (e.g. using the "recorder" plugin) and upload it?

**Comment by MaxKellermann** [comment]
And explain why you believe the conversion from 48000:f:2 to 48000:32:2 is "meaningless". I say it is necessary, until you prove otherwise, or until you explain how you want it.

**Comment by janui** [comment]
Hello Max, Thanks for the fast responce. It is possible that the padded bits when converting from f to 32 are not being initialised correctly. Seems only to happen with mp3 (and mp2) formats when these are being cast via upmpdcli (http://www.lesbonscomptes.com/upmpdcli/) and the output is 32bit. Could be that this issue is related to upmpdcli  and not MPD. I will try to provide a recording for you. janui

**Comment by MaxKellermann** [comment]
I found a float-to-32 conversion bug, and wrote a fix. Not sure if this is really your problem. Try the fix, and if your problem persists, upload the recording.

**Comment by klassikmann** [comment]
I'm working with janui on this issue and retried the tests with mpd-0.20.23 plus upmpdcli and ffmpeg 4.0.2. In the full logs attached you find the analysis of eac3to and mediainfo (from a stream previously recorded) plus the (verbose) mpd logs.

All audio streams are within a MPEG2 transport stream container (mpeg2ts). Below you find the relevant parts (my assessment) of the logs.

**Test 1 OE1DD**
TS, 1 audio track, 0:00:10
1: AC3, German, 5.1 channels, 448kbps, 48kHz
This stream is played correctly, downmix to 48000:32:2 (for reference only).
Full log: [OE1DD.log](https://github.com/MusicPlayerDaemon/MPD/files/2537494/OE1DD.log)
Excerpt:
Oct 31 12:29 : decoder_thread: probing plugin ffmpeg
Oct 31 12:29 : ffmpeg: detected input format 'mpegts' (MPEG-TS (MPEG-2 Transport Stream))
...
Oct 31 12:29 : **ffmpeg: codec 'ac3'**
Oct 31 12:29 : decoder: audio_format=48000:f:6, seekable=false
Oct 31 12:29 : alsa_output: opened hw:0,0 type=HW
Oct 31 12:29 : alsa_output: buffer: size=8..65536 time=166..1365334
Oct 31 12:29 : alsa_output: period: size=4..32768 time=83..682667
Oct 31 12:29 : alsa_output: default period_time = buffer_time/4 = 500000/4 = 125000
Oct 31 12:29 : alsa_output: format=S32_LE (Signed 32 bit Little Endian)
Oct 31 12:29 : alsa_output: buffer_size=24000 period_size=6000
Oct 31 12:29 : output: opened plugin=alsa name="snd_rpi_hifiberry_digi" audio_format=48000:32:2
Oct 31 12:29 : **output: converting in=48000:f:6 -> f=48000:f:6 -> out=48000:32:2**

**Test 2 OE1**
TS, 1 audio track, 0:00:12
1: MP2, German, 2.0 channels, 256kbps, 48kHz
Sound output: silence
Full Log: 
[OE1.log](https://github.com/MusicPlayerDaemon/MPD/files/2537502/OE1.log)
Excerpt:
Oct 31 12:28 : decoder_thread: probing plugin ffmpeg
Oct 31 12:28 : ffmpeg: detected input format 'mpegts' (MPEG-TS (MPEG-2 Transport Stream)
...
Oct 31 12:28 : **ffmpeg: codec 'mp2'**
Oct 31 12:28 : decoder: audio_format=48000:f:2, seekable=false
Oct 31 12:28 : alsa_output: opened hw:0,0 type=HW
Oct 31 12:28 : alsa_output: buffer: size=8..65536 time=166..1365334
Oct 31 12:28 : alsa_output: period: size=4..32768 time=83..682667
Oct 31 12:28 : alsa_output: default period_time = buffer_time/4 = 500000/4 = 125000
Oct 31 12:28 : ffmpeg/mpegts: pid=1a5 pes_code=0x1c0
Oct 31 12:28 : alsa_output: format=S32_LE (Signed 32 bit Little Endian)
Oct 31 12:28 : alsa_output: buffer_size=24000 period_size=6000
Oct 31 12:28 : output: opened plugin=alsa name="snd_rpi_hifiberry_digi" audio_format=48000:32:2
Oct 31 12:28 : **output: converting in=48000:f:2 -> f=48000:f:2 -> out=48000:32:2**
...
Oct 31 12:28 : **player: Decoder is too slow; playing silence to avoid xrun** (5x)

**Test 3 hr1**
TS, 1 audio track, 0:00:17
1: MP2, German, 2.0 channels, 320kbps, 48kHz
Sound output: silence plus occasional cracks
Full log: 
[hr1.log](https://github.com/MusicPlayerDaemon/MPD/files/2537503/hr1.log)
Excerpt:
Oct 31 12:29 : decoder_thread: probing plugin ffmpeg
Oct 31 12:29 : ffmpeg: detected input format 'mpegts' (MPEG-TS (MPEG-2 Transport Stream))
...
Oct 31 12:29 : **ffmpeg: codec 'mp2'**
Oct 31 12:29 : decoder: audio_format=48000:f:2, seekable=false
Oct 31 12:29 : alsa_output: opened hw:0,0 type=HW
Oct 31 12:29 : alsa_output: buffer: size=8..65536 time=166..1365334
Oct 31 12:29 : alsa_output: period: size=4..32768 time=83..682667
Oct 31 12:29 : alsa_output: default period_time = buffer_time/4 = 500000/4 = 125000
Oct 31 12:29 : alsa_output: format=S32_LE (Signed 32 bit Little Endian)
Oct 31 12:29 : alsa_output: buffer_size=24000 period_size=6000
Oct 31 12:29 : output: opened plugin=alsa name="snd_rpi_hifiberry_digi" audio_format=48000:32:2
Oct 31 12:29 : **output: converting in=48000:f:2 -> f=48000:f:2 -> out=48000:32:2**

**Test 4 hr2-kultur**
TS, 2 audio tracks, 0:00:17
1: MP2, German, 2.0 channels, 320kbps, 48kHz, 22ms
2: AC3, German, 2.0 channels, 448kbps, 48kHz
Sound output: silence plus occasional cracks
Full log: 
[hr2-kultur.log](https://github.com/MusicPlayerDaemon/MPD/files/2537504/hr2-kultur.log)
Excerpt:
Oct 31 12:30 : decoder_thread: probing plugin ffmpeg
Oct 31 12:30 : ffmpeg: detected input format 'mpegts' (MPEG-TS (MPEG-2 Transport Stream))
...
Oct 31 12:30 : **ffmpeg: codec 'mp2'**
Oct 31 12:30 : decoder: audio_format=48000:f:2, seekable=false
Oct 31 12:30 : alsa_output: opened hw:0,0 type=HW
Oct 31 12:30 : alsa_output: buffer: size=8..65536 time=166..1365334
Oct 31 12:30 : alsa_output: period: size=4..32768 time=83..682667
Oct 31 12:30 : alsa_output: default period_time = buffer_time/4 = 500000/4 = 125000
Oct 31 12:30 : alsa_output: format=S32_LE (Signed 32 bit Little Endian)
Oct 31 12:30 : alsa_output: buffer_size=24000 period_size=6000
Oct 31 12:30 : output: opened plugin=alsa name="snd_rpi_hifiberry_digi" audio_format=48000:32:2
Oct 31 12:30 : **output: converting in=48000:f:2 -> f=48000:f:2 -> out=48000:32:2**
...
Oct 31 12:30 : **player: Decoder is too slow; playing silence to avoid xrun**

**Comment by janui** [comment]
Hello Max,
Now working with klassikmann on this issue. Could you reopen it? Or do you want to handle it as a new issue?

**Comment by MaxKellermann** [comment]
If my patch hasn't fixed the issue, then let's continue here.

**Comment by klassikmann** [comment]
I used 0.20.23 in my comment above; the patch didn't fix the issue.
0.21 was too brand-new; I will re-test with 0.21 early next week.

**Comment by MaxKellermann** [comment]
"Test 1 OE1DD" proves that the float to 32 bit conversion does work correctly, doesn't it?

Is this bug report about "cracks" and "silence" or about some undefined "noise"?

Can you eliminate upmpdcli from the equation, and given me the original stream URLs to reproduce the problem? upmpdcli may insert timing problems into the whole, and cause those "cracks" by letting MPD's buffer starve. I don't know, I don't have upmpdcli and thus am unable to support it.

**Comment by klassikmann** [comment]
> "Test 1 OE1DD" proves that the float to 32 bit conversion does work correctly, doesn't it?

Yes. I left the test case OE1DD to demonstrate that for AC3 5.1 "converting in=48000:f:6 -> f=48000:f:6 -> out=48000:32:2" works but not the conversion from MP2 ("converting in=48000:f:2 -> f=48000:f:2 -> out=48000:32:2").

> Is this bug report about "cracks" and "silence" or about some undefined "noise"?

It's about "cracks" and "silence". The original post issued by janui ist based on my earlier tests, and I wrote to him erroneously about "noise" instead of "cracks" and "silence".

> Can you eliminate upmpdcli from the equation, and given me the original stream URLs to reproduce the problem?

I prepared a short (19 sec) snippet of the stream hr1 (Test 3 above). It's a MPEG2 transport stream containing one audio stream (MP2, German, 2.0 channels, 320kbps, 48kHz). Due to requirements of the bug tracking system I encapsulated the .ts in a .zip: [test.zip](https://github.com/MusicPlayerDaemon/MPD/files/2545954/test.zip)
The corresponding mpd.log is rather similar to my earlier logs: [mpd.log](https://github.com/MusicPlayerDaemon/MPD/files/2545957/mpd.log)
